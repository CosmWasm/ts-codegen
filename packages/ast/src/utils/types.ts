import * as t from '@babel/types';
import { camel, pascal } from 'case';
import { propertySignature } from './babel';
import { RenderContext } from '../context';
import { JSONSchema } from '@cosmology/ts-codegen-types';

export function getResponseType(
  context: RenderContext,
  underscoreName: string
) {
  const methodName = camel(underscoreName);
  return pascal(
    context.contract?.responses?.[underscoreName]?.title
    ??
    // after v1.1 is adopted, we can deprecate this and require the above response
    `${methodName}Response`
  );
};

const getTypeStrFromRef = ($ref: string): string => {
  if ($ref?.startsWith('#/definitions/')) {
    return $ref.replace('#/definitions/', '');
  }
  throw new Error('what is $ref: ' + $ref);
}

export const getTypeFromRef = ($ref: string) => {
  return t.tsTypeReference(t.identifier(getTypeStrFromRef($ref)))
}

const getArrayTypeFromRef = ($ref: string) => {
  return t.tsArrayType(
    getTypeFromRef($ref)
  );
}


export const getTypeOrRef = (obj: JSONSchema): t.TSType => {
  if (obj.$ref) {
    return getTypeFromRef(obj.$ref);
  } else if (typeof obj.type === 'string') {
    return getType(obj.type);
  } else if (Array.isArray(obj.type)) {
    // Handle array of types by creating a union type
    console.log(obj.type);
    const types = obj.type.map(t => getType(t));
    return t.tsUnionType(types);
  } else {
    throw new Error('Type definition not supported or missing: ' + JSON.stringify(obj));
  }
};

const getArrayTypeFromItems = (
  context: RenderContext,
  items: JSONSchema | JSONSchema[]
): t.TSArrayType => {
  if (Array.isArray(items)) {
    // if (context.options.types.itemsUseTuples === true) {
    // Using the modified method with tsTupleType
    // return t.tsArrayType(
    // t.tsTupleType(items.map(item => getTypeOrRef(item)))
    // );
    // } else {
    // Using the original method with nested arrays
    if (items.length > 0 && items[0].type === 'array') {
      return getArrayTypeFromItems(context, items[0]);
    }
    return items.length > 0 ? t.tsArrayType(t.tsArrayType(getTypeOrRef(items[0]))) : t.tsArrayType(t.tsArrayType(t.tsAnyKeyword()));
    // }
  } else {
    const detect = detectType(items.type);

    if (detect.type === 'array') {
      if (Array.isArray(items.items)) {
        return t.tsArrayType(
          t.tsArrayType(
            getTypeOrRef(items.items[0])
          )
        );
      } else if (items.items) {
        return t.tsArrayType(
          getArrayTypeFromItems(context, items.items)
        );
      } else {
        // Handle the case where items.items is undefined for array types
        return t.tsArrayType(t.tsAnyKeyword())
      }
    }

    return t.tsArrayType(
      getType(detect.type)
    );
  }
};

export const detectType = (type: string | string[]) => {
  let optional = false;
  let theType = '';
  if (Array.isArray(type)) {
    if (type.length !== 2) {
      throw new Error('[getType(array length)] case not handled by transpiler. contact maintainers.')
    }
    const [nullableType, nullType] = type;
    if (nullType !== 'null') {
      throw new Error('[getType(null)] case not handled by transpiler. contact maintainers.')
    }
    theType = nullableType;
    optional = true;
  } else {
    theType = type;
  }

  return {
    type: theType,
    optional
  };
}

export const getTypeInfo = (
  context: RenderContext,
  info: JSONSchema
) => {
  let type = undefined;
  let optional = undefined;

  // Handle nullable types using 'anyOf' schema
  if (Array.isArray(info.anyOf)) {
    if (info.anyOf.length !== 2) {
      throw new Error('case not handled by transpiler. contact maintainers.');
    }
    const [nullableType, nullType] = info.anyOf;
    if (nullType?.type !== 'null') {
      throw new Error('[nullableType.type]: case not handled by transpiler. contact maintainers.');
    }
    if (!nullableType?.$ref) {
      if (nullableType.title) {
        type = t.tsTypeReference(t.identifier(nullableType.title));
      } else {
        throw new Error('[nullableType.title] case not handled by transpiler. contact maintainers.');
      }
    } else {
      type = getTypeFromRef(nullableType.$ref);
    }
    optional = true;
  } else if (typeof info.type === 'string') {
    console.log('is string')
    // Process type when it's explicitly defined
    if (info.type === 'array') {
      if (typeof info.items === 'object' && !Array.isArray(info.items)) {
        if (info.items.$ref) {
          type = getArrayTypeFromRef(info.items.$ref);
        } else if (info.items.title) {
          type = t.tsArrayType(t.tsTypeReference(t.identifier(info.items.title)));
        } else if (info.items.type) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
        // type = getArrayTypeFromItems(context, info.items);
      } else {
        if (Array.isArray(info.items)) {
          type = getArrayTypeFromItems(context, info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
        }
      }
    } else {
      const detect = detectType(info.type);
      type = getType(detect.type);
      optional = detect.optional;
    }
  } else if (Array.isArray(info.type)) {
    console.log('is arr', info.type)
    // Handle multiple types, typically nullable types
    if (info.type.length !== 2) {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }
    const [nullableType, nullType] = info.type;
    if (nullType !== 'null') {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2));
    }

    if (nullableType === 'array' && typeof info.items === 'object' && !Array.isArray(info.items)) {
      if (info.items.type) {
        const detect = detectType(info.items.type);
        if (detect.type === 'array') {
          console.log('items type')
          type = t.tsArrayType(getArrayTypeFromItems(context, info.items));
        } else {
          type = t.tsArrayType(getType(detect.type));
        }
        optional = detect.optional;
      } else if (info.items.$ref) {
        type = getArrayTypeFromRef(info.items.$ref);
      } else if (info.items.type) {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        throw new Error('[info.items] case not handled by transpiler. contact maintainers.');
      }
    } else {
      const detect = detectType(nullableType);
      optional = detect.optional;
      if (detect.type === 'array') {
        type = getArrayTypeFromItems(context, info.items);
      } else {
        type = getType(detect.type);
      }
    }

    optional = true;  // Ensure optional is set for nullable types
  }
  console.log({ type })

  return {
    type,
    optional
  };
};

export const getType = (type: string) => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'boolean':
      return t.tSBooleanKeyword();
    case 'integer':
      return t.tsNumberKeyword();
    case 'null':
      return t.tsNullKeyword();
    case 'number':
      return t.tsNumberKeyword();
    default:
      throw new Error('contact maintainers [unknown type]: ' + type);
  }
}

export const getPropertyType = (
  context: RenderContext,
  schema: JSONSchema,
  prop: string
) => {
  const props = schema.properties ?? {};
  let info = props[prop];

  let type = null;
  let optional = !schema.required?.includes(prop);

  if (info.allOf && info.allOf.length === 1) {
    info = info.allOf[0];
  }

  if (typeof info.$ref === 'string') {
    type = getTypeFromRef(info.$ref)
  }

  const typeInfo = getTypeInfo(context, info);
  if (typeof typeInfo.optional !== 'undefined') {
    optional = typeInfo.optional;
  }
  if (typeof typeInfo.type !== 'undefined') {
    console.log(JSON.stringify(typeInfo, null, 2))
    type = typeInfo.type;
  }

  if (!type) {
    throw new Error('cannot find type for ' + JSON.stringify(info))
  }

  if (schema.required?.includes(prop)) {
    optional = false;
  }

  return { type, optional };
};


export function getPropertySignatureFromProp(
  context: RenderContext,
  jsonschema: JSONSchema,
  prop: string,
  camelize: boolean
) {
  if (jsonschema.properties[prop].type === 'object') {
    if (jsonschema.properties[prop].title) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(t.identifier(jsonschema.properties[prop].title))
        )
      );
    } else {
      throw new Error('getPropertySignatureFromProp() contact maintainer');
    }
  }

  if (Array.isArray(jsonschema.properties[prop].allOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].allOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    // @ts-ignore:next-line
    const uniqUnionTypes = [...new Set(unionTypes)];

    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }
  } else if (Array.isArray(jsonschema.properties[prop].oneOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].oneOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    }).filter(a => typeof a === 'string');
    // ONLY SUPPORT string types for now!
    // e.g. NOT supporting nullable types or type: string[] yet

    const uniqUnionTypes = [...new Set(unionTypes)];
    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0] as string)
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ as string)
              )
            )
          )
        ),
        isOptional
      );
    }

  }

  try {
    getPropertyType(context, jsonschema, prop);
  } catch (e) {
    console.log(e);
    console.log(JSON.stringify(jsonschema, null, 2), prop);
  }

  const { type, optional } = getPropertyType(context, jsonschema, prop);
  return propertySignature(
    camelize ? camel(prop) : prop,
    t.tsTypeAnnotation(
      type
    ),
    optional
  );
}

export const getParamsTypeAnnotation = (
  context: RenderContext,
  jsonschema: any,
  camelize: boolean = true
): t.TSTypeAnnotation => {
  const keys = Object.keys(jsonschema.properties ?? {});

  if (!keys.length && jsonschema.$ref) {
    return t.tsTypeAnnotation(getTypeFromRef(jsonschema.$ref))
  }

  if (!keys.length) return undefined;

  const typedParams = keys.map(prop => getPropertySignatureFromProp(
    context,
    jsonschema,
    prop,
    camelize
  ));

  return t.tsTypeAnnotation(
    t.tsTypeLiteral(
      // @ts-ignore:next-line
      [
        ...typedParams
      ]
    )
  )
}

export const createTypedObjectParams = (
  context: RenderContext,
  jsonschema: JSONSchema,
  camelize: boolean = true
): (t.Identifier | t.Pattern | t.RestElement) => {

  const keys = Object.keys(jsonschema.properties ?? {});
  if (!keys.length) {
    // is there a ref?
    if (jsonschema.$ref) {
      const obj = context.refLookup(jsonschema.$ref);
      // If there is a oneOf, then we need to create a type for it
      if (obj?.oneOf) {
        // the actual type of the ref
        const refType = jsonschema.$ref.split('/').pop();
        const refName = camel(refType);
        const id = t.identifier(refName);
        id.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(refType)));
        return id
      } else if (obj) {
        return createTypedObjectParams(
          context,
          obj,
          camelize
        );
      }
    }

    // no results...
    return;
  }

  const params = keys.map(prop => {
    return t.objectProperty(
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      false,
      true
    );
  });

  const obj = t.objectPattern(
    [
      ...params
    ]
  );

  obj.typeAnnotation = getParamsTypeAnnotation(
    context,
    jsonschema,
    camelize
  );

  return obj;
};