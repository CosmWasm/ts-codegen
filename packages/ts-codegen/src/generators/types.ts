import { pascal } from "case";
import { header } from '../utils/header';
import { join } from "path";
import { sync as mkdirp } from "mkdirp";
import * as t from '@babel/types';
import { writeFileSync } from 'fs';
import generate from "@babel/generator";
import { clean } from "../utils/clean";
import { findAndParseTypes, findExecuteMsg, findQueryMsg } from '../utils';
import { ContractInfo, RenderContext, TSTypesOptions } from "wasm-ast-types";
import { BuilderFile } from "../builder";

export default async (
  name: string,
  contractInfo: ContractInfo,
  outPath: string,
  tsTypesOptions?: TSTypesOptions
): Promise<BuilderFile[]> => {

  const { schemas } = contractInfo;
  const context = new RenderContext(contractInfo, {
    types: tsTypesOptions ?? {}
  });
  const options = context.options.types;

  const localname = pascal(name) + '.types.ts';
  const ExecuteMsg = findExecuteMsg(schemas);
  const typeHash = await findAndParseTypes(schemas);

  const body = [];

  // TYPES
  Object.values(typeHash).forEach((type: t.Node) => {
    body.push(
      clean(type)
    )
  });

  // alias the ExecuteMsg (deprecated option)
  if (options.aliasExecuteMsg && ExecuteMsg) {
    body.push(
      t.exportNamedDeclaration(
        t.tsTypeAliasDeclaration(
          t.identifier(`${name}ExecuteMsg`),
          null,
          t.tsTypeReference(t.identifier('ExecuteMsg'))
        )
      )
    );
  }

  function addEntryPointAlias(msgName: string) {
    body.push(
      t.exportNamedDeclaration(
        t.tsTypeAliasDeclaration(
          t.identifier(`${name}${msgName}`),
          null,
          t.tsTypeReference(t.identifier(msgName))
        )
      )
    );
  }

  if (options.aliasEntryPoints) {
    if (ExecuteMsg) {
      addEntryPointAlias('ExecuteMsg');
    }
    if (findQueryMsg(schemas)) {
      addEntryPointAlias('QueryMsg');
    }
  }


  const imports = context.getImports();
  const code = header + generate(
    // @ts-ignore
    t.program([
      ...imports,
      ...body
    ])
  ).code;

  mkdirp(outPath);
  const filename = join(outPath, localname);
  writeFileSync(filename, code);

  return [
    {
      type: 'type',
      contract: name,
      localname,
      filename,
    }
  ]
};
