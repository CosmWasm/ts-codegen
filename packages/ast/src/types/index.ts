export * from './context';
export * from './plugins';