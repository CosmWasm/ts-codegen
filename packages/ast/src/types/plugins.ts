export interface ReactQueryOptions {
  enabled?: boolean;
  optionalClient?: boolean;
  version?: 'v3' | 'v4';
  mutations?: boolean;
  camelize?: boolean;
  queryKeys?: boolean
  queryFactory?: boolean
}

export interface TSClientOptions {
  enabled?: boolean;
  execExtendsQuery?: boolean;
  noImplicitOverride?: boolean;
}
export interface MessageComposerOptions {
  enabled?: boolean;
}
export interface MessageBuilderOptions {
  enabled?: boolean;
}
export interface RecoilOptions {
  enabled?: boolean;
}
export interface TSTypesOptions {
  enabled?: boolean;
  itemsUseTuples?: boolean;
  // deprecated
  aliasExecuteMsg?: boolean;
  aliasEntryPoints?: boolean;
}

export interface UseContractsOptions {
  enabled?: boolean;
};
