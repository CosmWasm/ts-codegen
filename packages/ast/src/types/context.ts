import { IDLObject, JSONSchema } from "@cosmwasm/ts-codegen-types";
import { MessageBuilderOptions, MessageComposerOptions, ReactQueryOptions, RecoilOptions, TSClientOptions, TSTypesOptions, UseContractsOptions } from "./plugins";

export interface ContractInfo {
  schemas: JSONSchema[];
  responses?: Record<string, JSONSchema>;
  idlObject?: Partial<IDLObject>;
};
export interface RenderOptions {
  enabled?: boolean;
  types?: TSTypesOptions;
  recoil?: RecoilOptions;
  messageComposer?: MessageComposerOptions;
  messageBuilder?: MessageBuilderOptions;
  client?: TSClientOptions;
  reactQuery?: ReactQueryOptions;
  useContractsHook?: UseContractsOptions;
}

export interface ProviderInfo {
  classname: string,
  filename: string,
  basename: string,
}
