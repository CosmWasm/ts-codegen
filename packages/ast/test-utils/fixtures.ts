// @ts-nocheck
import { JSONSchema } from '../src/types';

import canExecuteRelayResponseData from '../../../__fixtures__/vectis/govec/can_execute_relay_response.json';
import cosmosMsgForEmptyData from '../../../__fixtures__/vectis/govec/cosmos_msg_for__empty.json';
import executeMsgForEmptyData from '../../../__fixtures__/vectis/govec/execute_msg_for__empty.json';
import infoResponseData from '../../../__fixtures__/vectis/govec/info_response.json';
import relayTransactionData from '../../../__fixtures__/vectis/govec/relay_transaction.json';

import nftIdlVersionData from '../../../__fixtures__/idl-version/accounts-nft/account-nft.json';

export const nftIdlVersion: JSONSchema = nftIdlVersionData;
export const vectisCanExecuteRelayResponse: JSONSchema = canExecuteRelayResponseData;
export const vectisCosmosMsgForEmpty: JSONSchema = cosmosMsgForEmptyData;
export const vectisExecuteMsgForEmpty: JSONSchema = executeMsgForEmptyData;
export const vectisInfoResponse: JSONSchema = infoResponseData;
export const vectisRelayTransaction: JSONSchema = relayTransactionData;
