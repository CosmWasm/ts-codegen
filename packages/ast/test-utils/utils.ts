import generate from '@babel/generator';
import { RenderOptions } from '@cosmwasm/ts-codegen-ast/types';
import { ExecuteMsg, IDLObject, JSONSchema, QueryMsg } from '@cosmwasm/ts-codegen-types';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { join } from 'path';

import { RenderContext } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
};

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
};

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options);
};

interface GlobContractLegacy {
  name: `/${string}.json`;
  content: JSONSchema;
}

interface GlobContract {
  name: `/${string}.json`;
  content: IDLObject;
}

const fixtureDir = join(__dirname, '../../../__fixtures__');
const globCache: Record<string, GlobContractLegacy[] | GlobContract[]> = {};

export const globIdlBasedContracts = (p: string): GlobContract[] => {
  if (globCache[p]) return globCache[p] as GlobContract[];
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const globLegacyContracts = (p: string): GlobContractLegacy[] => {
  if (globCache[p]) return globCache[p] as GlobContractLegacy[];
  // @ts-ignore
  const contracts: GlobContractLegacy[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const getMsgExecuteLegacyFixture = (
  p: string,
  name: `/${string}.json`
): ExecuteMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as ExecuteMsg;
};

export const getMsgQueryLegacyFixture = (
  p: string,
  name: `/${string}.json`
): QueryMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as QueryMsg;
};

export const getLegacyFixture = (
  p: string,
  name: `/${string}.json`
): JSONSchema => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as JSONSchema;
};
