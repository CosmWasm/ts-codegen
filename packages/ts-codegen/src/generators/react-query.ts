import { pascal } from 'case';
import { header } from '../utils/header';
import { join } from 'path';
import { sync as mkdirp } from 'mkdirp';
import * as w from 'wasm-ast-types';
import { RenderContext } from 'wasm-ast-types';
import * as t from '@babel/types';
import { writeFileSync } from 'fs';
import generate from '@babel/generator';
import { findAndParseTypes, findExecuteMsg, findQueryMsg } from '../utils';
import {
  getMessageProperties,
  ReactQueryOptions,
  ContractInfo
} from 'wasm-ast-types';
import { BuilderFile } from '../builder';

export default async (
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  reactQueryOptions?: ReactQueryOptions
): Promise<BuilderFile[]> => {
  const { schemas } = contractInfo;
  const context = new RenderContext(contractInfo, {
    reactQuery: reactQueryOptions ?? {}
  });
  const options = context.options.reactQuery;

  const localname = pascal(`${contractName}`) + '.react-query.ts';
  const ContractFile =
    pascal(`${contractName}`) + `client`;
  const TypesFile = pascal(`${contractName}`) + '.types';

  const QueryMsg = findQueryMsg(schemas);
  const ExecuteMsg = findExecuteMsg(schemas);
  const typeHash = await findAndParseTypes(schemas);

  const isAbstractApp = context.options.abstractApp?.enabled;

  const ExecuteClient = pascal(
    `${contractName}${isAbstractApp ? 'App' : ''}Client`
  );
  const QueryClient = pascal(
    `${contractName}${isAbstractApp ? 'App' : ''}QueryClient`
  );

  const body = [];

  const clientImports = [];

  if (QueryMsg) {
    clientImports.push(QueryClient);
  }

  // check that there are commands within the exec msg
  const shouldGenerateMutationHooks =
    ExecuteMsg &&
    options?.version === 'v4' &&
    options?.mutations &&
    getMessageProperties(ExecuteMsg).length > 0;

  if (shouldGenerateMutationHooks) {
    clientImports.push(ExecuteClient);
  }

  // general contract imports
  body.push(w.importStmt(Object.keys(typeHash), `./${TypesFile}`));

  // client imports
  body.push(w.importStmt(clientImports, `./${ContractFile}`));

  // query messages
  if (QueryMsg) {
    [].push.apply(
      body,
      w.createReactQueryHooks({
        context,
        queryMsg: QueryMsg,
        contractName: contractName,
        QueryClient
      })
    );
  }

  if (shouldGenerateMutationHooks) {
    [].push.apply(
      body,
      w.createReactQueryMutationHooks({
        context,
        execMsg: ExecuteMsg,
        contractName: contractName,
        ExecuteClient
      })
    );
  }

  if (typeHash.hasOwnProperty('Coin')) {
    // @ts-ignore
    delete context.utils.Coin;
  }
  const imports = context.getImports();
  const code = header + generate(t.program([...imports, ...body])).code;

  mkdirp(outPath);
  writeFileSync(join(outPath, localname), code);

  return [
    {
      type: 'react-query',
      contract: contractName,
      localname,
      filename: join(outPath, localname)
    }
  ];
};
