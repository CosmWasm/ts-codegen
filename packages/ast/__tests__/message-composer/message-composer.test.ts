import execute_msg from "../../../../__fixtures__/basic/execute_msg_for__empty.json";
import ownership from "../../../../__fixtures__/basic/ownership.json";

import {
  createMessageComposerClass,
  createMessageComposerInterface,
} from "./message-composer";
import { expectCode, makeContext } from "../../test-utils";
import * as t from "@babel/types";
import { createReactQueryMutationHooks } from "../react-query";

it("execute classes", () => {
  const ctx = makeContext(execute_msg);
  expectCode(
    createMessageComposerClass(
      ctx,
      "SG721MessageComposer",
      "SG721Message",
      execute_msg
    )
  );
});

it("createMessageComposerInterface", () => {
  const ctx = makeContext(execute_msg);
  expectCode(createMessageComposerInterface(ctx, "SG721Message", execute_msg));
});

it("ownershipClass", () => {
  const ctx = makeContext(ownership);
  expectCode(
    createMessageComposerClass(
      ctx,
      "OwnershipMessageComposer",
      "OwnershipMessage",
      ownership
    )
  );
});

it("ownershipInterface", () => {
  const ownershipCtx = makeContext(ownership);
  expectCode(
    createMessageComposerInterface(ownershipCtx, "OwnershipMessage", ownership)
  );
});
