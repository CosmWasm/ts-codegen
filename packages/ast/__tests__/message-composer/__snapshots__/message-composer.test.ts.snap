// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMessageComposerInterface 1`] = `
"export interface SG721Message {
  contractAddress: string;
  sender: string;
  transferNft: ({
    recipient,
    tokenId
  }: {
    recipient: string;
    tokenId: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  sendNft: ({
    contract,
    msg,
    tokenId
  }: {
    contract: string;
    msg: Binary;
    tokenId: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  approve: ({
    expires,
    spender,
    tokenId
  }: {
    expires?: Expiration;
    spender: string;
    tokenId: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  revoke: ({
    spender,
    tokenId
  }: {
    spender: string;
    tokenId: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  approveAll: ({
    expires,
    operator
  }: {
    expires?: Expiration;
    operator: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  revokeAll: ({
    operator
  }: {
    operator: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  mint: ({
    extension,
    owner,
    tokenId,
    tokenUri
  }: {
    extension: Empty;
    owner: string;
    tokenId: string;
    tokenUri?: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  burn: ({
    tokenId
  }: {
    tokenId: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
}"
`;

exports[`execute classes 1`] = `
"export class SG721MessageComposer implements SG721Message {
  sender: string;
  contractAddress: string;
  constructor(sender: string, contractAddress: string) {
    this.sender = sender;
    this.contractAddress = contractAddress;
    this.transferNft = this.transferNft.bind(this);
    this.sendNft = this.sendNft.bind(this);
    this.approve = this.approve.bind(this);
    this.revoke = this.revoke.bind(this);
    this.approveAll = this.approveAll.bind(this);
    this.revokeAll = this.revokeAll.bind(this);
    this.mint = this.mint.bind(this);
    this.burn = this.burn.bind(this);
  }
  transferNft = ({
    recipient,
    tokenId
  }: {
    recipient: string;
    tokenId: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          transfer_nft: {
            recipient,
            token_id: tokenId
          }
        })),
        funds: funds_
      })
    };
  };
  sendNft = ({
    contract,
    msg,
    tokenId
  }: {
    contract: string;
    msg: Binary;
    tokenId: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          send_nft: {
            contract,
            msg,
            token_id: tokenId
          }
        })),
        funds: funds_
      })
    };
  };
  approve = ({
    expires,
    spender,
    tokenId
  }: {
    expires?: Expiration;
    spender: string;
    tokenId: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          approve: {
            expires,
            spender,
            token_id: tokenId
          }
        })),
        funds: funds_
      })
    };
  };
  revoke = ({
    spender,
    tokenId
  }: {
    spender: string;
    tokenId: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          revoke: {
            spender,
            token_id: tokenId
          }
        })),
        funds: funds_
      })
    };
  };
  approveAll = ({
    expires,
    operator
  }: {
    expires?: Expiration;
    operator: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          approve_all: {
            expires,
            operator
          }
        })),
        funds: funds_
      })
    };
  };
  revokeAll = ({
    operator
  }: {
    operator: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          revoke_all: {
            operator
          }
        })),
        funds: funds_
      })
    };
  };
  mint = ({
    extension,
    owner,
    tokenId,
    tokenUri
  }: {
    extension: Empty;
    owner: string;
    tokenId: string;
    tokenUri?: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          mint: {
            extension,
            owner,
            token_id: tokenId,
            token_uri: tokenUri
          }
        })),
        funds: funds_
      })
    };
  };
  burn = ({
    tokenId
  }: {
    tokenId: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          burn: {
            token_id: tokenId
          }
        })),
        funds: funds_
      })
    };
  };
}"
`;

exports[`ownershipClass 1`] = `
"export class OwnershipMessageComposer implements OwnershipMessage {
  sender: string;
  contractAddress: string;
  constructor(sender: string, contractAddress: string) {
    this.sender = sender;
    this.contractAddress = contractAddress;
    this.setFactory = this.setFactory.bind(this);
    this.updateOwnership = this.updateOwnership.bind(this);
  }
  setFactory = ({
    newFactory
  }: {
    newFactory: string;
  }, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          set_factory: {
            new_factory: newFactory
          }
        })),
        funds: funds_
      })
    };
  };
  updateOwnership = (action: Action, funds_?: Coin[]): MsgExecuteContractEncodeObject => {
    return {
      typeUrl: "/cosmwasm.wasm.v1.MsgExecuteContract",
      value: MsgExecuteContract.fromPartial({
        sender: this.sender,
        contract: this.contractAddress,
        msg: toUtf8(JSON.stringify({
          update_ownership: action
        })),
        funds: funds_
      })
    };
  };
}"
`;

exports[`ownershipInterface 1`] = `
"export interface OwnershipMessage {
  contractAddress: string;
  sender: string;
  setFactory: ({
    newFactory
  }: {
    newFactory: string;
  }, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
  updateOwnership: (action: Action, funds_?: Coin[]) => MsgExecuteContractEncodeObject;
}"
`;
